mailbox/event/event-rabbitmq/src/main/java/org/apache/james/mailbox/events/ReactorRabbitMQChannelPool.java

```diff
@@ -0,0 +1,130 @@
+import org.apache.commons.pool2.impl.DefaultPooledObject;
+import org.apache.commons.pool2.impl.GenericObjectPool;
+import org.apache.commons.pool2.impl.GenericObjectPoolConfig;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.rabbitmq.client.Channel;
+import com.rabbitmq.client.Connection;
+
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.SignalType;
+import reactor.core.scheduler.Schedulers;
+import reactor.rabbitmq.ChannelPool;
+
+class ReactorRabbitMQChannelPool implements ChannelPool {
+
+    static class ChannelFactory extends BasePooledObjectFactory<Channel> {
+
+        private static final Logger LOGGER = LoggerFactory.getLogger(ChannelFactory.class);
+
+        private static final int MAX_RETRIES = 5;
+        private static final Duration RETRY_FIRST_BACK_OFF = Duration.ofMillis(100);
+        private static final Duration MAX_BACK_OFF = Duration.ofMillis(Long.MAX_VALUE);
+
+        private final Mono<Connection> connectionMono;
+
+        ChannelFactory(Mono<Connection> connectionMono) {
+            this.connectionMono = connectionMono;
+        }
+
+        @Override
+        public Channel create() throws Exception {
+            return connectionMono
+                .flatMap(this::openChannel)
+                .block();
+        }
+
+        private Mono<Channel> openChannel(Connection connection) {
+            return Mono.fromCallable(connection::openChannel)
+                .map(maybeChannel ->
+                    maybeChannel.orElseThrow(() -> new RuntimeException("RabbitMQ reached to maximum opened channels, cannot get more channels")))
+                .retryBackoff(MAX_RETRIES, RETRY_FIRST_BACK_OFF, MAX_BACK_OFF, Schedulers.elastic())
+                .doOnError(throwable -> LOGGER.error("error when creating new channel", throwable));
+        }
+
+        @Override
+        public PooledObject<Channel> wrap(Channel obj) {
+            return new DefaultPooledObject<>(obj);
+        }
+
+        @Override
+        public void destroyObject(PooledObject<Channel> pooledObject) throws Exception {
+            Channel channel = pooledObject.getObject();
+            if (channel.isOpen()) {
+                channel.close();
+            }
+        }
+    }
+
+    private static final long MAXIMUM_BORROW_TIMEOUT_IN_MS = Duration.ofSeconds(5).toMillis();
+
+    private final GenericObjectPool<Channel> pool;
+    private final ChannelFactory channelFactory;
+
+    ReactorRabbitMQChannelPool(Mono<Connection> connectionMono, int poolSize) {
+        this.channelFactory = new ChannelFactory(connectionMono);
+
+        GenericObjectPoolConfig<Channel> config = new GenericObjectPoolConfig<>();
+        config.setMaxTotal(poolSize);
+        this.pool = new GenericObjectPool<>(channelFactory, config);
+    }
+
+    @Override
+    public Mono<? extends Channel> getChannelMono() {
+        return Mono.fromCallable(() -> pool.borrowObject(MAXIMUM_BORROW_TIMEOUT_IN_MS));
```

Comment:
what happens after 5 seconds?

Reply:
after 5 seconds, pool.borrowObject() throws an timeout exception. and this method return an error Mono

Reply:
That probably a good solution. maybe we should extract the borrow call into a tryGet method or something that make it clear it can fail.